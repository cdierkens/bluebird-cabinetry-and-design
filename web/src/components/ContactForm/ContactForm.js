import axios from "axios";
import { Field, Form, Formik } from "formik";
import { Link } from "gatsby";
import React, { useState } from "react";
import * as Yup from "yup";
import Alert from "../Alert";
import Button from "../Button";
import Container from "../container";
import Grid from "../Grid";
import Input from "../Input";
import Select from "../Select";

const phoneRegExp = /^\({0,1}\d{3}[-|)]{0,1}\d{3}-{0,1}\d{4}$/;

const contactSchema = Yup.object().shape({
  name: Yup.string().required("Please enter your name."),
  phone: Yup.string().matches(
    phoneRegExp,
    "Please enter a 10 digit phone number"
  ),
  email: Yup.string()
    .required("Please enter your email.")
    .email("Please enter a valid email."),
  message: Yup.string()
    .required("Please enter a message.")
    .min(25, "Please provide more details."),
});

const ContactForm = () => {
  const [response, setResponse] = useState({
    status: "init",
    message: "",
  });

  return (
    <div className="bg-gray-light py-12">
      {response.status !== "init" && response.message && (
        <Alert
          status={response.status}
          onClose={() => setResponse({ status: "init", message: "" })}
        >
          {response.message}
        </Alert>
      )}

      <Container>
        <Grid>
          <p className="col-span-3 sm:col-span-6 md:col-start-2 text-center text-blue-dark">
            Please take a minute to tell us more about yourself and your
            project. We will be in contact soon!
          </p>

          <Formik
            initialValues={{
              name: "",
              phone: "",
              email: "",
              budget: "",
              interest: "",
              message: "",
              username: "",
            }}
            validationSchema={contactSchema}
            onSubmit={async (values, { resetForm }) => {
              try {
                const { data } = await axios.post(
                  "/.netlify/functions/send-contact-email",
                  values
                );
                resetForm();

                setResponse({
                  status: "success",
                  message: data,
                });
              } catch (error) {
                if (error.response) {
                  const { data } = error.response;

                  if (typeof data === "string") {
                    setResponse({
                      status: "error",
                      message: data,
                    });
                  } else {
                    setResponse({
                      status: "error",
                      message: error.message,
                    });
                  }
                } else {
                  console.error(error);
                }
              }
            }}
          >
            {() => (
              <Form className="block col-span-3 sm:col-span-6 md:col-start-2">
                <Input placeholder="Jane Doe" label="Name*" name="name" />

                <Input
                  placeholder="555-555-5555"
                  label="Phone Number"
                  name="phone"
                />

                <Input
                  placeholder="jane@example.com"
                  label="Email*"
                  name="email"
                />

                <Select label="Budget" name="budget">
                  <Select.Option value="" disabled>
                    Please select a budget...
                  </Select.Option>
                  <Select.Option value="$0 to $10,000">
                    $0 to $10,000
                  </Select.Option>
                  <Select.Option value="$11,000 to $25,000">
                    $11,000 to $25,000
                  </Select.Option>
                  <Select.Option value="$26,000 to $50,000">
                    $26,000 to $50,000
                  </Select.Option>
                  <Select.Option value="$51,000 to $75,000">
                    $51,000 to $75,000
                  </Select.Option>
                  <Select.Option value="$76,000 to $100,000">
                    $76,000 to $100,000
                  </Select.Option>
                  <Select.Option value="$100,000 plus">
                    $100,000 plus
                  </Select.Option>
                </Select>

                <Input
                  placeholder="cabinetry, remodeling, renovations, etc"
                  label="Areas of Interest"
                  name="interest"
                />

                <Input type="textarea" label="Message*" name="message" />

                <Field name="username" value="some value" className="sr-only" />

                <p className="-mt-2 ml-2 text-sm text-blue-dark">
                  Fields marked with * are required.
                </p>

                <div className="flex items-center justify-center mt-4">
                  <Button variant="blue" type="submit">
                    Send
                  </Button>
                </div>
                <div className="text-center text-sm text-blue-dark p-8">
                  If you are a trade professional, see our
                  <Link to="/trade-services"> To The Trade Services </Link>
                  page.
                </div>
              </Form>
            )}
          </Formik>
        </Grid>
      </Container>
    </div>
  );
};

export default ContactForm;
