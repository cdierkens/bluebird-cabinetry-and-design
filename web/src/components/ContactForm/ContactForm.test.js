import { waitFor } from "@testing-library/dom";
import { fireEvent, render } from "@testing-library/react";
import React from "react";
import ContactForm from "./ContactForm";

describe("<ContactForm />", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<ContactForm />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should show an error message when name is left blank", async () => {
    // Arrange
    const { getByLabelText, getByText } = render(<ContactForm />);
    const input = getByLabelText("Name*");

    // Act
    fireEvent.blur(input);

    // Assert
    await waitFor(() => {
      expect(getByText("Please enter your name.")).toBeInTheDocument();
    });
  });

  it("should show an error message phone number is invalid", async () => {
    // Arrange
    const { getByLabelText, getByText } = render(<ContactForm />);
    const input = getByLabelText("Phone Number");

    // Act
    fireEvent.input(input, { target: { value: "abcdefghij" } });
    fireEvent.blur(input);

    // Assert
    await waitFor(() => {
      expect(
        getByText("Please enter a 10 digit phone number")
      ).toBeInTheDocument();
    });
  });
});
