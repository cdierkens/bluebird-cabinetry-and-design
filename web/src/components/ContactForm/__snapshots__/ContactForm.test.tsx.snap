// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContactForm /> matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Container bg-gray-light py-12 my-12"
  >
    <div
      class="Grid gap-3"
    >
      <p
        class="col-span-3 sm:col-span-6 md:col-start-2 text-center text-blue-dark"
      >
        Please take a minute to tell us more about yourself and your project. We will be in contact soon!
      </p>
      <form
        action="#"
        class="block col-span-3 sm:col-span-6 md:col-start-2"
      >
        <label
          class="block relative pb-5 mb-2"
        >
          <span
            class="block text-blue-dark text-base ml-1 mb-1"
          >
            Name*
          </span>
          <input
            class="block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-base placeholder-gray-darker focus:outline-none focus:ring shadow-sm"
            name="name"
            placeholder="Jane Doe"
            type="text"
            value=""
          />
        </label>
        <label
          class="block relative pb-5 mb-2"
        >
          <span
            class="block text-blue-dark text-base ml-1 mb-1"
          >
            Phone Number
          </span>
          <input
            class="block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-base placeholder-gray-darker focus:outline-none focus:ring shadow-sm"
            name="phone"
            placeholder="555-555-5555"
            type="text"
            value=""
          />
        </label>
        <label
          class="block relative pb-5 mb-2"
        >
          <span
            class="block text-blue-dark text-base ml-1 mb-1"
          >
            Email*
          </span>
          <input
            class="block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-base placeholder-gray-darker focus:outline-none focus:ring shadow-sm"
            name="email"
            placeholder="jane@example.com"
            type="text"
            value=""
          />
        </label>
        <label
          class="SelectWrapper"
        >
          <span
            class="block text-blue-dark text-base ml-1 mb-1"
          >
            Budget
          </span>
          <div
            class="relative"
          >
            <select
              class="Select"
              name="budget"
            >
              <option
                class="Option"
                disabled=""
                value=""
              >
                Please select a budget...
              </option>
              <option
                class="Option"
                value="$0 to $10,000"
              >
                $0 to $10,000
              </option>
              <option
                class="Option"
                value="$11,000 to $25,000"
              >
                $11,000 to $25,000
              </option>
              <option
                class="Option"
                value="$26,000 to $50,000"
              >
                $26,000 to $50,000
              </option>
              <option
                class="Option"
                value="$51,000 to $75,000"
              >
                $51,000 to $75,000
              </option>
              <option
                class="Option"
                value="$76,000 to $100,000"
              >
                $76,000 to $100,000
              </option>
              <option
                class="Option"
                value="$100,000 plus"
              >
                $100,000 plus
              </option>
            </select>
            <div
              class="Arrow"
            >
              <svg
                class="fill-current h-4 w-4"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
                />
              </svg>
            </div>
          </div>
        </label>
        <label
          class="block relative pb-5 mb-2"
        >
          <span
            class="block text-blue-dark text-base ml-1 mb-1"
          >
            Areas of Interest
          </span>
          <input
            class="block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-base placeholder-gray-darker focus:outline-none focus:ring shadow-sm"
            name="interest"
            placeholder="cabinetry, remodeling, renovations, etc"
            type="text"
            value=""
          />
        </label>
        <label
          class="block relative pb-5 mb-2"
        >
          <span
            class="block text-blue-dark text-base ml-1 mb-1"
          >
            Message*
          </span>
          <textarea
            class="block w-full bg-white border border-gray-light px-3 py-3 pr-8 leading-tight font-sans text-base placeholder-gray-darker focus:outline-none focus:ring shadow-sm"
            name="message"
            rows="6"
          />
        </label>
        <input
          class="sr-only"
          name="username"
          value="some value"
        />
        <p
          class="-mt-2 ml-2 text-sm text-blue-dark"
        >
          Fields marked with * are required.
        </p>
        <div
          class="flex items-center justify-center mt-4"
        >
          <button
            class="ButtonBlue focus:outline-none focus:ring"
            type="submit"
          >
            Send
          </button>
        </div>
        <div
          class="text-center text-sm text-blue-dark p-8"
        >
          If you are a trade professional, see our
          <a
            href="/trade-services"
          >
             To The Trade Services 
          </a>
          page.
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
